### PR TITLE
feat: add tests section in doc editor

### PR DESCRIPTION
**Pull request by saravmajestic**
### Problem
- As dbt power user, i want to manage tests for my model using vscode-dbt-power-user extension

### Solution
- Added `tests` section in documentation editor panel
- User can add generic tests, update `relationships` and `accepted_values` test values, view custom tests
- This new tests section will show up only if `dbt.enableTests` is enabled in config

### Screenshot/Demo
image
image
image
image

### How to test
- Install extension with this current change
- load a dbt project
- open a model in editor
- open `documentation editor` panel from bottom
- enable `new experience` if not enabled
- click `Tests` button
- click + button to add new tests
- once added click `save documentation` to save the updates
- click existing tests to delete or view more details in right side
- reload the vscode to make sure the changes are intact

## Checklist

- [x] I have run this code and it appears to resolve the stated issue
- [x] I have added the necessary tests
- [ ] `README.md` updated and added information about my change
- [ ] Documentation website is updated with information related to this change

### CODE REVIEW

File: src/domain.ts [hunk 1]
```diff
@@ -62,13 +62,30 @@ interface DocMetaData {
...
+}
+
+export interface TestMetaData {
   path: string;
   database: string;
   schema: string;
   alias: string;
   raw_sql: string;
   column_name?: string;
+  test_metadata?: {
```
Comment (saravmajestic) [inline src/domain.ts:84]: Added missing field for tests data

File: webview_panels/src/testUtils/documentation/index.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import {
+  DBTDocumentation,
+  DBTDocumentationColumn,
+  Source,
+  DBTModelTest,
+  DbtGenericTests,
+  TestMetadataAcceptedValuesKwArgs,
+} from "@modules/documentationEditor/state/types";
+
+const getRandomNumber = (maximum = 10, minimum = 5) =>
```
Comment (saravmajestic) [inline webview_panels/src/testUtils/documentation/index.ts:12]: Sample docs data for storybook

File: src/webview_provider/datapilotPanel.ts [hunk 3]
```diff
@@ -43,13 +45,10 @@ export class DataPilotPanel extends AltimateWebviewProvider {
...
 
-    commands.registerCommand(
-      "dbtPowerUser.resetDatapilot",
-      () =>
-        this._panel?.webview.postMessage({
-          command: "datapilot:reset",
-          args: {},
-        }),
+    commands.registerCommand("dbtPowerUser.resetDatapilot", () =>
+      this.sendResponseToWebview({
```
Comment (saravmajestic) [inline src/webview_provider/datapilotPanel.ts:49]: used a common method `sendResponseToWebview` instead of `this.panel.webview.postMessage`

File: webview_panels/src/modules/documentationEditor/components/settings/DocGeneratorSettings.tsx [hunk 4]
```diff
@@ -39,47 +36,45 @@ const DocGeneratorSettings = (): JSX.Element => {
...
-                  onChange={handleChange}
-                  options={Persona.map((l) => ({ label: l, value: l }))}
-                />
-              </Label>
-            </Stack>
-          </Stack>
-        </RightSidePanel>
-      ) : null}
-    </>
+    <Drawer
```
Comment (saravmajestic) [inline webview_panels/src/modules/documentationEditor/components/settings/DocGeneratorSettings.tsx:39]: Used common `Drawer` component to handle functions like keyboard escape, clicking on backdrop etc

File: package.json [hunk 5]
```diff
@@ -72,7 +72,12 @@
           },
           "dbt.enableNewDocsPanel": {
             "type": "boolean",
-            "description": "Enable the new docs panel in dbt."
+            "description": "Enable the new docs panel in dbt.",
+            "default": true
+          },
+          "dbt.enableTests": {
+            "type": "boolean",
+            "description": "Enable the new tests experience in documentation editor."
```
Comment (mdesmet) [inline package.json:80]: I don't think this is required as we already have a button to show/display tests.

File: src/domain.ts [hunk 6]
```diff
@@ -62,13 +62,30 @@ interface DocMetaData {
   character: number;
 }
 
-interface TestMetaData {
+interface TestMetadataKwArgs {
```
Comment (mdesmet) [inline src/domain.ts:65]: I think you can do this through a union type and create multiple types per variant

Comment (mdesmet) [inline src/domain.ts:65]: I think the naming can be better, for example `TestSpecification`

File: src/services/dbtProjectService.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
+import { Uri, window } from "vscode";
+import { DBTProject } from "../manifest/dbtProject";
+import { DBTProjectContainer } from "../manifest/dbtProjectContainer";
+import { ManifestCacheProjectAddedEvent } from "../manifest/event/manifestCacheChangedEvent";
+import { provideSingleton } from "../utils";
+
+@provideSingleton(DbtProjectService)
+export class DbtProjectService {
```
Comment (mdesmet) [inline src/services/dbtProjectService.ts:8]: Naming can be improved. This class is more of a wrapper around the manifest, to make it easier to query the manifest. Also why you don't let it manage the eventMap updates as well? That way we can reuse it in a lot more places.

Just search for `Map<string, ManifestCacheProjectAddedEvent>` and you see what I mean.

File: src/services/docGenService.ts [hunk 8]
```diff
@@ -65,7 +67,7 @@ export class DocGenService {
       }
       const enableNewDocsPanel = workspace
         .getConfiguration("dbt")
-        .get<boolean>("enableNewDocsPanel", false);
```
Comment (mdesmet) [inline src/services/docGenService.ts:68]: I thought we agreed to get rid of this configuration

File: src/services/docGenService.ts [hunk 9]
```diff
@@ -228,6 +222,36 @@ export class DocGenService {
...
+      return;
+    }
+
+    const event = eventMap.get(projectRootpath.fsPath);
+    if (event === undefined) {
+      this.dbtTerminal.debug("no event for project: ", projectRootpath.fsPath);
+      return;
+    }
+    return event;
+  }
```
Comment (mdesmet) [inline src/services/docGenService.ts:253]: Good candidate to move to dbtProjectService.

File: src/webview_provider/docsEditPanel.ts [hunk 10]
```diff
@@ -249,6 +253,74 @@ export class DocsEditViewPanel implements WebviewViewProvider {
     await this.resolveWebviewView(this.panel!, this.context!, this.token!);
   };
 
+  private getTestDataByColumn(message: any, columnName: string) {
+    const tests = message.tests as undefined | TestMetaData[];
+
+    if (!tests?.length) {
+      this.terminal.debug("tests", "No test data passed");
```
Comment (mdesmet) [inline src/webview_provider/docsEditPanel.ts:260]: "tests" may look a bit generic if seeing that in the logs. Can you put the class name instead?

File: src/webview_provider/docsEditPanel.ts [hunk 11]
```diff
@@ -249,6 +253,74 @@ export class DocsEditViewPanel implements WebviewViewProvider {
...
+
+        // set values if test is accepted_values
+        if (test.test_metadata?.name === "accepted_values") {
+          return {
+            accepted_values: { values: test.test_metadata.kwargs.values },
+          };
+        }
+        return test.test_metadata?.name;
+      })
+      // Filter duplicate values
```
Comment (mdesmet) [inline src/webview_provider/docsEditPanel.ts:291]: WDYM with duplicate values?

File: src/services/docGenService.ts (hunk 8 above)
Comment (saravmajestic) [reply src/services/docGenService.ts:68]: We will remove this config in next release. in the first version, we will make new doc gen as default and remove switching button. but user can still go to settings and enable legacy. In next version we will remove this config

File: package.json (hunk 5 above)
Comment (saravmajestic) [reply package.json:80]: In the first release, we dont want to show tests by default. Once we introduce infer tests from dbt, we will enable it by default

File: src/webview_provider/docsEditPanel.ts (hunk 11 above)
Comment (saravmajestic) [reply src/webview_provider/docsEditPanel.ts:291]: In some edge cases, relationships are set twice, though it is not visible in UI. This reduce block is to cleanup any duplicate tests added, since we are dealing with generic tests.

File: src/domain.ts (hunk 6 above)
Comment (saravmajestic) [reply src/domain.ts:65]: yes right, I did that in UI code, but didnt check this. Will fix this

Comment (saravmajestic) [reply src/domain.ts]: fixed

Comment (saravmajestic) [reply src/domain.ts]: fixed

File: src/services/dbtProjectService.ts (hunk 7 above)
Comment (saravmajestic) [reply src/services/dbtProjectService.ts:12]: https://github.com/AltimateAI/vscode-dbt-power-user/pull/910/commits/3de02a01bbc569f55351cd9d55964b6aac698e8f Modified and cleaned up some existing code.

File: src/services/docGenService.ts (hunk 9 above)
Comment (saravmajestic) [reply src/services/docGenService.ts]: moved

File: src/webview_provider/docsEditPanel.ts (hunk 10 above)
Comment (saravmajestic) [reply src/webview_provider/docsEditPanel.ts]: fixed

File: src/services/dbtProjectService.ts (hunk 7 above)
Comment (mdesmet) [reply src/services/dbtProjectService.ts:12]: Can you also rename this to `QueryManifestService` (also the filename)?

File: src/services/docGenService.ts [hunk 12]
```diff
@@ -174,27 +175,14 @@ export class DocGenService {
...
     }
 
     const currentFilePath = window.activeTextEditor.document.uri;
-    const project = this.getProject();
-    if (project === undefined) {
-      return undefined;
-    }
-    const event = eventMap.get(project.projectRoot.fsPath);
+
+    const event = this.dbtProjectService.getEventByCurrentProject();
```
Comment (mdesmet) [inline src/services/docGenService.ts:185]: Maybe safer to pass down the url of the document to this method.

**Review by mdesmet — COMMENTED** [review]
Please restrict the number of usages `activeTextEditor.document.uri`, I usually prefer to do this only once, it is probably even better to use the one from the event. Otherwise you will introduce some randomness in case a user quickly switches between windows.

File: src/services/docGenService.ts (hunk 12 above)
Comment (saravmajestic) [reply src/services/docGenService.ts]: fixed

File: src/manifest/dbtProject.ts [hunk 13]
```diff
@@ -518,6 +518,34 @@ export class DBTProject implements Disposable {
...
+      this.terminal.debug(
+        "getColumnValues",
+        "finding distinct values for column",
+        true,
+        { model, column },
+      );
+      const query = `select ${column} as values from ${model} group by ${column}`;
+      const queryExecution = await this.dbtProjectIntegration.executeSQL(
+        query,
+        1000,
```
Comment (saravmajestic) [inline src/manifest/dbtProject.ts:532]: executeSQL expects limit to be passed, i just set to 1000 for now. Is this ok? or should we make limit as optional in executeSql and ignore limit?

File: src/services/docGenService.ts [hunk 14]
```diff
@@ -453,10 +441,39 @@ export class DocGenService {
...
+    if (!eventResult?.event || !eventResult?.currentDocument) {
+      return undefined;
+    }
+
+    const {
+      event: { nodeMetaMap, graphMetaMap, testMetaMap },
+      currentDocument,
+    } = eventResult;
+    const tableName = path.basename(currentDocument.fileName, ".sql");
+    this.dbtTerminal.info(
```
Comment (mdesmet) [inline src/services/docGenService.ts:455]: info will pollute log

File: webview_panels/src/modules/documentationEditor/components/tests/DisplayTestDetails.tsx [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+
+  const [testCode, setTestCode] = useState("");
+  const [isInEditMode, setIsInEditMode] = useState(false);
+
+  const isEditableTest =
+    test.test_metadata?.name !== DbtGenericTests.NOT_NULL &&
+    test.test_metadata?.name !== DbtGenericTests.UNIQUE;
+
+  const handleDelete = () => {
+    panelLogger.info("delete test", test);
```
Comment (mdesmet) [inline webview_panels/src/modules/documentationEditor/components/tests/DisplayTestDetails.tsx:60]: will pollute log

File: src/domain.ts [hunk 16]
```diff
@@ -62,13 +62,36 @@ interface DocMetaData {
...
+// for accepted_values
+interface TestMetadataAcceptedValues {
+  values?: string[];
+}
+
+// for relationship
+interface TestMetadataRelationships {
+  field?: string;
+  to?: string;
+}
```
Comment (mdesmet) [inline src/domain.ts:81]: This is not a Union type but a single type.

File: src/manifest/dbtProject.ts (hunk 13 above)
Comment (mdesmet) [reply src/manifest/dbtProject.ts:532]: We need to set it much lower, set it to 100. We should probably add some wording around that.
The application will crash if we load too much data.

File: src/webview_provider/docsEditPanel.ts (hunk 11 above)
Comment (mdesmet) [reply src/webview_provider/docsEditPanel.ts:290]: This seems fixing symptoms rather than causes.

File: src/webview_provider/docsEditPanel.ts [hunk 17]
```diff
@@ -249,6 +249,83 @@ export class DocsEditViewPanel implements WebviewViewProvider {
...
+        }
+
+        // set values if test is accepted_values
+        if (test.test_metadata?.name === "accepted_values") {
+          return {
+            accepted_values: { values: test.test_metadata.kwargs.values },
+          };
+        }
+        return test.test_metadata?.name;
+      })
```
Comment (mdesmet) [inline src/webview_provider/docsEditPanel.ts:289]: We need to test if there are any custom properties that we don't support and bail out in that case.

**Review by mdesmet — COMMENTED** [review]
Logging parameters should be improved, tests is a little vague.

File: src/services/docGenService.ts (hunk 14 above)
Comment (saravmajestic) [reply src/services/docGenService.ts]: fixed

File: src/webview_provider/docsEditPanel.ts [hunk 18]
```diff
@@ -249,6 +249,83 @@ export class DocsEditViewPanel implements WebviewViewProvider {
...
+
+        if (test.relationships) {
+          acc["relationships"] = { relationships: test.relationships };
+          return acc;
+        }
+        return acc;
+      }, {});
+
+    const data = Object.values(filteredTests);
+    this.terminal.info(
```
[the remark this answers is not in the document]
Comment (saravmajestic) [reply src/webview_provider/docsEditPanel.ts]: fixed

File: src/domain.ts (hunk 16 above)
Comment (saravmajestic) [reply src/domain.ts]: fixed

File: src/webview_provider/newDocsGenPanel.ts [hunk 19]
```diff
@@ -115,10 +116,12 @@ export class NewDocsGenPanel
...
-        } catch (err) {
-          window.showErrorMessage(
-            extendErrorWithSupportLinks((err as Error).message),
-          );
+        } catch (error) {
+          const message =
+            error instanceof PythonException
+              ? error.exception.message
+              : (error as Error).message;
+          window.showErrorMessage(extendErrorWithSupportLinks(message));
```
[the remark this answers is not in the document]
Comment (saravmajestic) [reply src/webview_provider/newDocsGenPanel.ts:124]: it is already added in `getColumnsValues` implementation [here](https://github.com/AltimateAI/vscode-dbt-power-user/pull/910/files#diff-707d5997df2b8462df3c94a31f5f39412064245691e45bbeb48e9584162ed1e5R538)

File: src/webview_provider/newDocsGenPanel.ts [hunk 20]
```diff
@@ -154,10 +157,12 @@ export class NewDocsGenPanel
...
```
[the remark this answers is not in the document]
Comment (saravmajestic) [reply src/webview_provider/newDocsGenPanel.ts:165]: fixed

File: src/webview_provider/newDocsGenPanel.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (saravmajestic) [reply src/webview_provider/newDocsGenPanel.ts:132]: fixed

File: webview_panels/src/modules/documentationEditor/components/tests/utils.ts [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+
+  if (
+    Object.values(DbtGenericTests).includes(
+      test.test_metadata.name as DbtGenericTests,
+    )
+  ) {
+    return DbtTestTypes.GENERIC;
+  }
+
+  if (test.test_metadata.namespace) {
```
Comment (saravmajestic) [inline webview_panels/src/modules/documentationEditor/components/tests/utils.ts:30]: if test metadata has a namespace, it is identified as external package test

File: webview_panels/src/modules/documentationEditor/components/tests/DisplayTestDetails.tsx (hunk 15 above)
Comment (saravmajestic) [reply webview_panels/src/modules/documentationEditor/components/tests/DisplayTestDetails.tsx:63]: these are just console logs in webview only. it will not show in terminal

**Review by mdesmet — APPROVED** [review]
